libs/chonky/src/lib.rs

```diff
@@ -1,26 +1,134 @@
 #![doc = include_str!("../README.md")]
 
-/// Adds two numbers together
-///
-/// # Example
-///
-/// ```rust
-/// use chonky::add;
-///
-/// assert_eq!(add(1, 3), 4);
-/// ```
-#[must_use]
-pub const fn add(left: u64, right: u64) -> u64 {
-    left + right
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum ChonkyError {
+    #[error("parsing error in pdf")]
+    ReadPdf,
+    #[error("pdfium error")]
+    Pdfium,
+    #[error("write error to system")]
+    Write,
+    #[error("Issues with CLI input")]
+    Arguments,
+}
+
+pub mod pdf_segmentation {
+    use error_stack::{Report, ResultExt};
+    use image::DynamicImage;
+    use pdfium_render::prelude::*;
+
+    use crate::ChonkyError;
+
+    /// Function to read the pdf
+    ///
+    /// # Errors
+    ///
+    /// Will return `Err` if `filename` does not exist or the user does not have
+    /// permission to read it.
+    pub fn load_pdf<'a>(
+        pdfium: &'a Pdfium,
+        file_path: &str,
+    ) -> Result<PdfDocument<'a>, Report<ChonkyError>> {
+        pdfium
+            .load_pdf_from_file(file_path, None)
+            .map_err(|err| Report::new(err).change_context(ChonkyError::Pdfium))
+    }
+
+    /// Takes in a pdf document and returns a vector list where each page
+    /// is processed into a raw image that can be later converted to any image format
+    ///
+    /// # Errors
+    /// Return an `PdfiumError::ImageError` if there was a problem with the image processing
+    /// operation, occurs when image cannot be encoded into specific image format
+    pub fn pdf_to_images(pdf: &PdfDocument) -> Result<Vec<DynamicImage>, Report<ChonkyError>> {
+        let mut images: Vec<DynamicImage> = Vec::new();
+
+        for page in pdf.pages().iter() {
+            //may adjust resolution depending on need
+            let resolution_width = 1000;
+
+            // Render the entire page to an image
+            let rendered_page = page
+                .render_with_config(&PdfRenderConfig::new().set_target_width(resolution_width))
+                .change_context(ChonkyError::Pdfium)?; // Renders the page to a PdfBitmap
+
+            // Convert PdfBitmap to DynamicImage
+            let dynamic_image = rendered_page.as_image();
```

Comment:
`as_image` is a bit deceptive here, as I've worked with `image` before, converting anything into an image object is fallible. Looking at the source code that is also the case here, but it will [simply panic](https://docs.rs/pdfium-render/latest/src/pdfium_render/pdf/bitmap.rs.html#312), as all methods used are public it might make sense to vendor this function in the future to remove the panicking code.

Reply:
I think vendoring in this function makes sense until they return `Result<DynamicImage, PdfiumError>`, I agree. Good spot, thank you!